internal/builders/container/README.md

```diff
@@ -0,0 +1,261 @@
+  uses: slsa-framework/slsa-github-generator/.github/workflows/generator_container_slsa3.yml@main
+  with:
+    image: ${{ needs.build.outputs.tag }}
+    registry-username: ${{ github.actor }}
+    # TODO(https://github.com/slsa-framework/slsa-github-generator/issues/492): Remove after GA release.
+    compile-generator: true
+  secrets:
+    registry-password: ${{ secrets.GITHUB_TOKEN }}
+```
+
+Here's an example of what it might look like all together.
+
+```yaml
+jobs:
+  # This step builds our image, pushes it, and outputs the repo hash digest.
+  build:
+    permissions:
+      contents: read
+      packages: write
+    outputs:
+      tag: ${{ steps.hash.outputs.tag }}
+      sha: ${{ steps.hash.outputs.sha }}
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout the repository
+        uses: actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b # v2.3.4
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@dc7b9719a96d48369863986a06765841d7ea23f6 # v2.0.0
+
+      - name: Authenticate Docker
+        uses: docker/login-action@49ed152c8eca782a232dede0303416e8f356c37b # v2.0.0
+        with:
+          registry: ${{ env.IMAGE_REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Extract metadata (tags, labels) for Docker
+        id: meta
+        uses: docker/metadata-action@69f6fc9d46f2f8bf0d5491e4aabe0bb8c6a4678a # v4.0.1
+        with:
+          images: ${{ env.IMAGE_REGISTRY }}/${{ env.IMAGE_NAME }}
+
+      - name: Build and push Docker image
+        uses: docker/build-push-action@e551b19e49efd4e98792db7592c17c09b89db8d8 # v3.0.0
+        id: build
+        with:
+          push: true
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
+
+      - name: Output digest
+        id: hash
+        env:
+          IMAGE_TAGS: ${{ steps.meta.outputs.tags }}
+          SHA: ${{ steps.build.outputs.digest }}
+        run: |
+          # sha output of docker/build-push-action is of the form 'sha256:<digest>'
+          sha=$(echo $SHA | cut -d':' -f2)
+          # docker/build-push-action outputs multiple tags including latest.
+          # It seems to put latest at the end of the list.
+          tag=$(echo "$IMAGE_TAGS" | head -n1)
+          echo "::set-output name=tag::$tag"
+          echo "::set-output name=sha::$sha"
+
+  # This step calls the container workflow to generate provenance and push it to
+  # the container registry.
+  provenance:
+    needs: [build]
+    permissions:
+      actions: read
+      id-token: write
+      contents: read
+      packages: write
+    if: startsWith(github.ref, 'refs/tags/')
+    uses: slsa-framework/slsa-github-generator/.github/workflows/generator_container_slsa3.yml@main
+    with:
+      # XXX: should this use the sha instead of the tag?
+      image: ${{ needs.build.outputs.tag }}
```

Comment:
I suppose that the image tag could be overwritten between the time it's pushed to the registry and the time `cosign attest` is run. I wonder if the sha should be used instead? I would think given how the `cosign` CLI works that this would be a common issue, but I don't really see any guidance in the [cosign docs](https://docs.sigstore.dev/cosign/attestation). The examples just seem to run on `latest`.

Reply:
There was a discussion related to this on the slack channel. The TL;DR is to sign digests and not tags.
https://sigstore.slack.com/archives/C01PZKDL4DP/p1656926576991169

There is also an issue to discourage signing tags.
https://github.com/sigstore/cosign/issues/2047